modules/nlu-testing/src/tools/visualisation.ts

```diff
@@ -0,0 +1,137 @@
+      z: simMat.matrix,
+      type: 'heatmap'
+    }
+  ]
+
+  return plotlyMatrixData
+}
+
+export async function computeScatterEmbeddings(state: BotState, logger: sdk.Logger) {
+  const pcaTrain = new PCA(state.trainDatas.map(o => o.utt_emb))
+  const varianceTrain = pcaTrain.getExplainedVariance()
+  const first3DimsTrain = varianceTrain.slice(0, 3)
+  const totalVarianceTrain = _.round(_.sum(first3DimsTrain), 2)
+  logger.info(`Top 3 train variance ${first3DimsTrain.map(o => _.round(o, 2))} Accounting for ${totalVarianceTrain} % `)
+
+  const groupedIntentsTrain = _.groupBy(state.trainDatas, 'intent')
+
+  let traces = Object.entries(groupedIntentsTrain).map(([k, v]: [string, any[]], i) => {
+    return {
+      x: v.map(o => pcaTrain.predict([o.utt_emb]).get(0, 0)),
+      y: v.map(o => pcaTrain.predict([o.utt_emb]).get(0, 1)),
+      z: v.map(o => pcaTrain.predict([o.utt_emb]).get(0, 2)),
+      mode: 'markers',
+      type: 'scatter3d',
+      name: k,
+      text: v.map(o => o.utt),
+      marker: { size: 8, color: i }
+    }
+  })
+
+  if (state.testDatas.length) {
+    const pcaTest = new PCA(state.testDatas.map(o => o.utt_emb))
+    const varianceTest = pcaTest.getExplainedVariance()
+    const first3DimsTest = varianceTest.slice(0, 3)
+    const totalVarianceTest = _.round(_.sum(first3DimsTest), 2)
+    logger.info(`Top 3 train variance ${first3DimsTest.map(o => _.round(o, 2))} Accounting for ${totalVarianceTest} % `)
+
+    const groupedIntentsTest = _.groupBy(state.testDatas, 'intent')
+    const tracesTest = Object.entries(groupedIntentsTest).map(([k, v]: [string, any[]], i) => {
+      return {
+        x: v.map(o => pcaTest.predict([o.utt_emb]).get(0, 0)),
+        y: v.map(o => pcaTest.predict([o.utt_emb]).get(0, 1)),
+        z: v.map(o => pcaTest.predict([o.utt_emb]).get(0, 2)),
+        mode: 'markers',
+        type: 'scatter3d',
+        name: `${k}_test`,
+        text: v.map(o => o.utt),
+        marker: { size: 8, color: i }
+      }
+    })
+    traces = traces.concat(tracesTest)
+  }
+  return traces
+}
+
+function minimalDistance(a: number[], b: number[][], index: number): number {
+  let minIndex = undefined
+  let minDistance = Number.MAX_SAFE_INTEGER
+
+  b.map((point, i) => {
+    const dist = distance.euclidean(a, point)
+    if (dist < minDistance && i !== index) {
+      minIndex = i
+      minDistance = dist
+    }
+  })
+
+  return minDistance
+}
+
+export function computeOutliers(state: BotState) {
+  const intentsData = _.groupBy(state.trainDatas, 'intent')
+  const dbPerIntent = _.mapValues(intentsData, o => {
+    const embedArray = o.map(e => e.utt_emb)
+    const meanDist = o.reduce((sum, curr, index) => {
+      if (index < o.length - 1) {
+        return sum + minimalDistance(curr.utt_emb, embedArray, index)
+      }
+      return sum / o.length
+    }, 0)
+
+    const dbscan = new clustering.DBSCAN()
+    // parameters: neighborhood radius, number of points in neighborhood to form a cluster
```

Comment:
When typings are missing, don't hesitate to create them yourself. It will help future work here 

```js
interface DbScan {
  run: (neighboorhoodRadius: number, numberOfPoints: number) => number
}

const dbscan: DbScan = new clustering.DBSCAN()
```

Reply:
100% agree with @allardy here. 

@Ierezell, can you do this super fast? If not, merge this anyway, then we'll come back and add some more typings.

This really needs to merged.

Reply:
Indeed, I added run and noise typings.